### ISSUE TITLE
Kubelet not starting with CRI-O on BTRFS (could not find device)

### ISSUE THREAD

**Issue by lyind** [kubernetes/kubernetes#94335]
**What happened**:
Kubelet v1.19.0 doesn't start on nodes with CRI-O 1.17.2 (storage_driver="btrfs") with BTRFS as root filesystem.

Related/previous issues:
https://github.com/kubernetes/kubernetes/issues/47046
https://github.com/kubernetes/kubernetes/issues/65204

This error shows up in the journal:
```
07:42.061303    2703 kubelet.go:1296] Failed to start ContainerManager failed to get rootfs info: failed to get device for dir "/var/lib/kubelet": could not find device with major: 0, minor: 16 in cached partitions map
```
The `minor: 16` is referring to the btrfs root filesystem, mounted at the time.

`stat /`:
```
  File: /
  Size: 162       	Blocks: 32         IO Block: 4096   directory
Device: 10h/16d	Inode: 256         Links: 1
Access: (0755/drwxr-xr-x)  Uid: ( 1000/ UNKNOWN)   Gid: ( 1000/ UNKNOWN)
Access: 2020-08-29 21:00:48.000000000 +0000
Modify: 2020-08-10 12:05:17.000000000 +0000
Change: 2020-08-29 21:04:42.683000680 +0000
 Birth: -
```

`cat /proc/$(pidof kubelet)/mountinfo`:
```
12 18 0:11 / /sys rw,nosuid,nodev,noexec,relatime shared:7 - sysfs sysfs rw
13 18 0:12 / /proc rw,relatime shared:11 - proc proc rw
14 18 0:5 / /dev rw,nosuid,noexec,relatime shared:2 - devtmpfs udev rw,size=32894084k,nr_inodes=8223521,mode=755
15 14 0:13 / /dev/pts rw,nosuid,noexec,relatime shared:3 - devpts devpts rw,gid=5,mode=620,ptmxmode=000
16 18 0:14 / /run rw,nosuid,nodev,noexec,relatime shared:5 - tmpfs tmpfs rw,size=6586408k,mode=755
18 1 0:15 / / rw,noatime shared:1 - btrfs /dev/sda2 rw,degraded,ssd,discard=async,space_cache=v2,autodefrag,subvolid=5,subvol=/
17 14 0:17 / /dev/shm rw,nosuid,nodev shared:4 - tmpfs tmpfs rw
19 16 0:18 / /run/lock rw,nosuid,nodev,noexec,relatime shared:6 - tmpfs tmpfs rw,size=5120k
20 12 0:19 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:8 - cgroup2 cgroup2 rw,nsdelegate
21 12 0:20 / /sys/fs/pstore rw,nosuid,nodev,noexec,relatime shared:9 - pstore pstore rw
22 12 0:21 / /sys/fs/bpf rw,nosuid,nodev,noexec,relatime shared:10 - bpf none rw,mode=700
23 14 0:10 / /dev/mqueue rw,nosuid,nodev,noexec,relatime shared:12 - mqueue mqueue rw
246 16 0:24 / /run/user/0 rw,nosuid,nodev,relatime shared:148 - tmpfs tmpfs rw,size=6586404k,nr_inodes=1646601,mode=700
189 18 0:15 /var/lib/containers/storage/btrfs /var/lib/containers/storage/btrfs rw,noatime - btrfs /dev/sda2 rw,degraded,ssd,discard=async,space_cache=v2,autodefrag,subvolid=5,subvol=/
```

**What you expected to happen**:
Kubelet starts and detects the filesystem backing `/var/lib/kubelet`, even if it is BTRFS.

**How to reproduce it (as minimally and precisely as possible)**:
 * Install kubeadm/kubectl 1.19.0
 * Install CRI-O, I used 1.17.2 from:
   `deb [arch=amd64] http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/testing/Debian_Testing /`
 * Configure CRI-O to use `storage_driver = "btrfs"`
 * Perform "kubeadm init" creating a new cluster
 * Observe journal (check for kubelet issues)

**Anything else we need to know?**:

A possible workaround is to make sure a bind mount exists which allows kubelet's logic to find the backing fileystem. Eg. add the following fstab entry and then perform `mount /var/lib/kubelet`:
```
/var/lib/kubelet /var/lib/kubelet none defaults,bind,nofail 0 0
```
The resulting mount should look somewhat like this:
```
273 18 0:15 /var/lib/kubelet /var/lib/kubelet rw,noatime shared:1 - btrfs /dev/sda2 rw,degraded,ssd,discard=async,space_cache=v2,autodefrag,subvolid=5,subvol=/
```

The workaround has been tested and seems to make kubelet work as expected.

**Environment**:
- Kubernetes version: 1.19.0
- OS: Debian Bullseye (Testing)
- Kernel: Linux 5.8.5
- Others: The `podman-rootless` package has been installed

**Comment by neolit123** [comment]
cc @dashpole

**Comment by dvonessen** [comment]
I am encounter the same issue. @lyind thank you for your workaround. It seems to work.

**Comment by valentin2105** [comment]
I got the same issue with latest stable Docker on Debian (btrfs) and k8s v1.19.2.

**Comment by seanho00** [comment]
I confirm the same issue shows up with `/var/lib/kubelet` on btrfs (on LVM on LUKS, Debian buster), running k3s.  No btrfs subvols.

The workaround works for me; for k3s I also need to bind mount `/var/lib/rancher`.  I did not experience this issue with 1.18.

**Comment by ftvkyo** [comment]
This also reproduces with Docker runtime.
Workaround with bind mount helps, however `kubeadm reset` automatically unmounts it so don't get confused.

`/var/lib/kubelet` is on root that is mounted like this: `/dev/nvme0n1p2 on / type btrfs (rw,relatime,ssd,space_cache,subvolid=256,subvol=/root)`,
and docker is configured to use `overlay2`.

**Comment by aleiphoenix** [comment]
Reproduced with k8s 1.19.3 on Gentoo. The rootfs is btrfs, `/var/lib/kubelt` is on rootfs, docker is configured with storage drigver `overlay2` .

`kubeadm init` automatically mount `/var/lib/kubelt` and unmount it on `kubeadm reset`.

Also confirmed bind mount workaround helps.

**Comment by gjkim42** [comment]
I think almost anyone using BTRFS as the root filesystem has encountered this issue. 
If https://github.com/google/cadvisor/pull/2744 is merged and applied to the kubernetes, this can be fixed.

Please take a look at this [PR](https://github.com/google/cadvisor/pull/2744)?

**Comment by gjkim42** [comment]
The [fix](https://github.com/google/cadvisor/pull/2752) has been merged to google/cadvisor.
I think if Kubernetes v1.21 includes the fix by using google/cadvisor v0.39, the issue will be resolved.

**Comment by plevart** [comment]
Just a note that I too have stumbled on this issue after reconfiguring CRI-O storage to use btrfs driver pointing with /etc/containers/storage.conf graphroot directly to a BTRFS subvolume (the root filesystem is not BTRFS). After reading this issue I tried mounting the BTRFS subvolume explicitly on some mountpoint and pointed with graphroot to this explicit subvolume mountpoint which solved the problem. The problem seems to occur only when pointing to a subvolume within some parent subvolume as this is not a real mountpoint.

**Comment by mblumber1** [comment]
Ok, so I just want to make sure I understand. Currently when i try to run kubeadm to create a new cluster while using btrfs, it gives me 
`[ERROR SystemVerification]: unsupported graph driver: btrfs. `
With Kubernetes v1.21, that will no longer be the case?

**Comment by plevart** [comment]
In my experience (with CRI-O 1.19 and k8s 1.19), the btrfs driver (in /etc/containers/storage.conf) is supported, but the combination does not work correctly when graphroot is pointed to a btrfs subvolume within some parent subvolume (I don't have experience with graphroot poited to some subdirectory of a default subvolume and/or when default subvolume is also a root filesystem, just when pointed directly to a subvolume within some parent subvolume which is not a root filesystem). For example, when btrfs filesystem is mounted like the following (`/etc/fstab`):

```
UUID=b3a3e532-bcfe-418a-a306-977d01c4ace6 /btrfs btrfs defaults 0 0
```

...and when I then create a subvolume:

`btrfs subvolume create /btrfs/containers`

...and then use it as a graphroot (`/etc/containers/storage.conf`):

```
driver = "btrfs"
graphroot = "/btrfs/containers"
```

But when I explicitly mount the subvolume to some mountpoint (`/etc/fstab`):

```
UUID=b3a3e532-bcfe-418a-a306-977d01c4ace6 /btrfs btrfs defaults 0 0
UUID=b3a3e532-bcfe-418a-a306-977d01c4ace6 /mnt/containers btrfs subvol=/containers 0 0
```

...and then use the explicitly mounted path as the graphroot (`/etc/containers/storage.conf`):

```
driver = "btrfs"
graphroot = "/mnt/containers"
```

...then it works correctly without any problems. I guess some logic (either in CRI-O/containers or in kubelet) tries to find the "mountpoint" of the filesystem used for graphroot or something related and can't do that correctly when graphroot is located in a btrfs subvolume that is not explicitly mounted somewhere.

**Comment by plevart** [comment]
@mblumber1 Ah, I see what you meant by saying that `btrfs` is not supported in `kubeadm`. I haven't checked that since I created k8s clusted with default (overlay) driver and later migrated to `btrfs`. It is possible that `kubeadm` disallows using `btrfs` because of problems like the one described above. But in my experience, when a subvolume is mounted explicitly, the btrfs works as expected - at least I didn't have problems with it.

**Comment by arianvp** [comment]
> I think if Kubernetes v1.21 includes the fix by using google/cadvisor v0.39, the issue will be resolved.

Can the fix be backported to 1.19? 

`kubeadm` only supports updating one version at a time:

> Skipping MINOR versions when upgrading is unsupported

 (https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/)
 
So this leaves people on btrfs with `1.18` with no upgrade path to `1.21`  which is not really nice.

**Comment by gjkim42** [comment]
@bobbypage 
We've discussed this issue at https://github.com/google/cadvisor/pull/2752#issuecomment-745351932

Regarding https://github.com/kubernetes/kubernetes/issues/94335#issuecomment-753317107
What do you think about this?

> > I think if Kubernetes v1.21 includes the fix by using google/cadvisor v0.39, the issue will be resolved.
> 
> Can the fix be backported to 1.19?
> 
> `kubeadm` only supports updating one version at a time:
> 
> > Skipping MINOR versions when upgrading is unsupported
> 
> (https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/)
> 
> So this leaves people on btrfs with `1.18` with no upgrade path to `1.21` which is not really nice.

Thanks!

**Comment by warmwaffles** [comment]
> Maybe, there are quite a lot of people using btrfs than I expected.
> What do you think about this?

I am running up against this issue as well. I am stuck with using k8s on 1.18

**Comment by bobbypage** [comment]
@gjkim42
> Maybe, there are quite a lot of people using btrfs than I expected. What do you think about this?

It sounds like for full fix, we'll need to cherrypick https://github.com/google/cadvisor/pull/2752 to be included in k8s 1.19 and 1.20. To do that the following is needed:

1. Backport https://github.com/google/cadvisor/pull/2752 to cAdvisor v0.37 and v0.38 branches
    * k8s 1.19 uses cAdvisor v0.37.3 - https://github.com/kubernetes/kubernetes/blob/release-1.19/go.mod#L56
    * k8s 1.20 uses cAdvisor v0.38.6 - https://github.com/kubernetes/kubernetes/blob/release-1.20/go.mod#L56
2. Cut new cAdvisor patch release for v0.37 and v0.38
3. Create [cherrypick PRs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md) against k8s on 1.19 and 1.20 bumping to newly released cAdvisor (v0.37 / v0.38) with the fix

I'm happy to cut the new cAdvisor release from my side if you want to proceed and think this is critical enough bug that is worth cherrypicking.

**Comment by bobbypage** [comment]
New cAdvisor has been cut with the BTRFs fix:

* v0.37.4 - https://github.com/google/cadvisor/releases/tag/v0.37.4
* v0.38.7 - https://github.com/google/cadvisor/releases/tag/v0.38.7
